packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -1470,17 +1520,28 @@ class _RenderDecoration extends RenderBox {
 
     if (label != null) {
       final double labelX = _boxParentData(label!).offset.dx;
+      // the +1 shifts the range of x from (-1.0, 1.0) to (0.0, 2.0)
+      final double floatAlign = decoration.floatingLabelAlignment.x + 1;
+      final double floatWidth = _boxSize(label).width * _kFinalLabelScale;
+      // When floating label is centered, its x is relative to
+      // _BorderContainer's x and is independent of label's x.
       switch (textDirection) {
         case TextDirection.rtl:
-          decoration.borderGap!.start = labelX + label!.size.width;
+          decoration.borderGap!.start = lerpDouble(labelX + _boxSize(label).width,
```

Comment:
In case https://github.com/flutter/flutter/pull/92940 gets merged before this PR, we should make sure they work together correctly.

Reply:
Could you elaborate?

Reply:
We should make sure that the label still looks ok when it is using both this and that PR, like a centered label with zero border gap padding for example.

Reply:
I don't think they are related. This PR handles the position of start and end of the border, while #92940 fixes the math of how to draw border corners.

Reply:
I just wanted to make sure that, say, if this PR is used to center a floating label that is very long so that it overlaps the curvature of the top left and top right border corners, then the padding should still have the correct padding.

However, my PR has been reverted.  We'll probably move forward with this one first then, and afterwards I'll double check that everything is still ok on my PR.

Reply:
If we use `labelText`, the text just shortens with an ellipsis.
![labelText](https://user-images.githubusercontent.com/44755140/143126489-b5bc3550-f4a9-448a-a4d1-8641029f31e8.png)
In the case of `label`, the field constraints the widget.
![label](https://user-images.githubusercontent.com/44755140/143126414-d32f5bec-fd2f-47de-b817-65bbfada2369.png)